### ISSUE TITLE
PV USB

### ISSUE THREAD

**Issue by marmarek** [marmarek/test#164]
**Reported by marmarek on 17 Apr 2012 12:59 UTC**
PV USB with Linux 3.x works fine, but needs some work to integrate with Qubes:
1. xen-usbfront and xen-usbback drivers (patch from xen-devel with some minor modification works ok). It is ready on my devel machine.
2. Add support for PV USB to libxl (or stay with standalone python scripts).
2a. This includes some script in backend VM (in some/most cases not dom0) that bind USB device to usbback driver - /usr/lib/qubes/unbind_pci_device.sh equivalent.
3. Add Qubes tool to manage USB devices (qvm-block equivalent).

Migrated-From: https://wiki.qubes-os.org/ticket/531

**Comment by marmarek** [comment]
**Commented by marmarek on 3 Oct 2012 23:25 UTC**
Some related info: https://groups.google.com/group/qubes-devel/browse_thread/thread/e002ae940061d897

**Comment by marmarek** [comment]
**Commented by marmarek on 8 Nov 2012 03:00 UTC**
qvm-usb code done by Alexandre Bezroutchko merged into master branch.
Kernel frontend/backend drivers are still unstable, but this isn't the scope of this ticket.

**Comment by marmarek** [comment]
**Commented by joanna on 8 Feb 2013 12:53 UTC**
We're still waiting for a working pvusb backend...

**Comment by marmarek** [comment]
**Commented by marmarek on 24 Oct 2014 12:01 UTC**
USBIP in 3.17 kernel (uploaded to unstable repo) is moved out of "staging" tree. Its architecture/interface looks good in terms of usability in Qubes: network connection is established in userspace, then single file descriptor (socket) is passed to the kernel.
Based on that, I've managed to run it over qrexec (minor change in qrexec was required - to use sockets instead of pipes). Tested on USB stick and webcam - looks pretty good. Test on internal bluetooth adapter failed (oops in backend), but also had problems with this device without USBIP in the middle. All with UsbVM of course.

Some remaining problems:
1. Performance - in R2 qrexec is pretty slow, also produces a lot of CPU usage in dom0. Hopefully will work much better in R3. But still, even in R2 it was enough for webcam.
2. Connection workflow - VM-VM qrexec services are normally initiated by one VM to another. In qvm-usb we'd like to dom0 arrange the VM-VM connection. In R2 it is rather simple to work around, as the connection always goes through dom0, but in R3 it will be somehow more tricky - but still doable of course. Or perhaps we should implement it without qrexec, just plain vchan?
